x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java

```diff
@@ -1196,6 +1200,77 @@ protected Expression rule(Expression e) {
             return e.foldable() ? Literal.of(e) : e;
         }
     }
+    
+    static class LocalRelationCountFolding extends OptimizerRule<LogicalPlan> {
+
+        @Override
+        protected LogicalPlan rule(LogicalPlan p) {
+            if (Verifier.isLocalRelation(p)) {
+                // collect aggregate aliases, if any
+                List<Alias> aliases = new ArrayList<>();
+                if (p instanceof Filter) {
+                    Filter f = (Filter) p;
+                    if (f.child() instanceof Aggregate) {
+                        ((Aggregate) f.child()).aggregates().forEach(a -> {
+                            if (a instanceof Alias && ((Alias) a).child() instanceof Count) {
+                                aliases.add((Alias) a);
+                            }
+                        });
+                    }
+                }
+                
+                // replace all aliases or COUNTs in the aggregates with 1
+                LogicalPlan changed = p.transformDown(agg -> {
+                    List<NamedExpression> values = new ArrayList<>();
+                    for (NamedExpression a : agg.aggregates()) {
+                        if (a instanceof Count) {
+                            values.add(Literal.of(a.source(), 1L));
+                        } else if (a instanceof Alias && ((Alias) a).child() instanceof Count) {
+                            values.add(Literal.of(a.name(), Literal.of(Source.EMPTY, 1L)));
+                        } else {
+                            values.add(a);
+                        }
+                    }
+
+                    if (values.equals(agg.aggregates()) == false) {
+                        return new Aggregate(agg.source(), agg.child(), agg.groupings(), values);
+                    }
+                    return agg;
+                }, Aggregate.class);
+                
+                // locate and replace any COUNT aliases and COUNTs in HAVINGs with 1
+                changed = changed.transformDown(filter -> {
+                    Expression cond = filter.condition();
+                    // replace the COUNT(*) alias with 1
```

Comment:
The comment is a bit inaccurate as we also replace in general the `count(xxx)` and also not only its aliases.